File: test/language/union_types/cpp98/EmptyUnionTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "union_types/empty_union/EmptyUnion.h"

namespace union_types
{
namespace empty_union
{

TEST(EmptyUnionTest, emptyConstructor)
{
    EmptyUnion emptyUnion;
    ASSERT_EQ(EmptyUnion::CHOICE_UNDEFINED, emptyUnion.choiceTag());
    ASSERT_EQ(0, emptyUnion.bitSizeOf());
}

TEST(EmptyUnionTest, bitStreamReaderConstructor)
{
    zserio::BitStreamReader reader(NULL, 0);

    EmptyUnion emptyUnion(reader);
    ASSERT_EQ(EmptyUnion::CHOICE_UNDEFINED, emptyUnion.choiceTag());
    ASSERT_EQ(0, emptyUnion.bitSizeOf());
}

TEST(EmptyUnionTest, choiceTag)
{
    EmptyUnion emptyUnion;
    ASSERT_EQ(EmptyUnion::CHOICE_UNDEFINED, emptyUnion.choiceTag());
}

TEST(EmptyUnionTest, bitSizeOf)
{
    EmptyUnion emptyUnion;
    const size_t bitPosition = 1;
    ASSERT_EQ(0, emptyUnion.bitSizeOf(bitPosition));
}

TEST(EmptyUnionTest, initializeOffsets)
{
    const size_t bitPosition = 1;
    EmptyUnion emptyUnion;
    ASSERT_EQ(bitPosition, emptyUnion.initializeOffsets(bitPosition));
}

TEST(EmptyUnionTest, operatorEquality)
{
    EmptyUnion emptyUnion1;
    EmptyUnion emptyUnion2;
    ASSERT_TRUE(emptyUnion1 == emptyUnion2);
}

TEST(EmptyUnionTest, hashCode)
{
    EmptyUnion emptyUnion1;
    EmptyUnion emptyUnion2;
    ASSERT_EQ(emptyUnion1.hashCode(), emptyUnion2.hashCode());
}

TEST(EmptyUnionTest, read)
{
    zserio::BitStreamReader reader(NULL, 0);
    EmptyUnion emptyUnion;
    emptyUnion.read(reader);
    ASSERT_EQ(EmptyUnion::CHOICE_UNDEFINED, emptyUnion.choiceTag());
    ASSERT_EQ(0, emptyUnion.bitSizeOf());
}

TEST(EmptyUnionTest, write)
{
    EmptyUnion emptyUnion;

    zserio::BitStreamWriter writer;
    emptyUnion.write(writer);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);

    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    EmptyUnion readEmptyUnion(reader);
    ASSERT_TRUE(emptyUnion == readEmptyUnion);
}

} // namespace empty_union
} // namespace union_types

File: test/language/union_types/java/union_types/EmptyUnionTest.java
package union_types;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import union_types.empty_union.EmptyUnion;

public class EmptyUnionTest
{
    @Test
    public void emptyConstructor()
    {
        final EmptyUnion emptyUnion = new EmptyUnion();
        assertEquals(0, emptyUnion.bitSizeOf());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException
    {
        final BitStreamReader reader = new ByteArrayBitStreamReader(new byte[0]);

        final EmptyUnion emptyUnion = new EmptyUnion(reader);
        assertEquals(0, emptyUnion.bitSizeOf());
    }

    @Test
    public void bitSizeOf()
    {
        final int bitPosition = 1;
        final EmptyUnion emptyUnion = new EmptyUnion();
        assertEquals(0, emptyUnion.bitSizeOf(bitPosition));
    }

    @Test
    public void initializeOffsets()
    {
        final int bitPosition = 1;
        final EmptyUnion emptyUnion = new EmptyUnion();
        assertEquals(bitPosition, emptyUnion.initializeOffsets(bitPosition));
    }

    @Test
    public void equals()
    {
        final EmptyUnion emptyUnion1 = new EmptyUnion();
        final EmptyUnion emptyUnion2 = new EmptyUnion();
        assertTrue(emptyUnion1.equals(emptyUnion2));
    }

    @Test
    public void hashCodeMethod()
    {
        final EmptyUnion emptyUnion1 = new EmptyUnion();
        final EmptyUnion emptyUnion2 = new EmptyUnion();
        assertEquals(emptyUnion1.hashCode(), emptyUnion2.hashCode());

        // use hardcoded values to check that the hash code is stable
        assertEquals(850, emptyUnion1.hashCode());
    }

    @Test
    public void read() throws IOException
    {
        final BitStreamReader reader = new ByteArrayBitStreamReader(new byte[0], 0);

        final EmptyUnion emptyUnion = new EmptyUnion();
        emptyUnion.read(reader);
        assertEquals(0, emptyUnion.bitSizeOf());
    }

    @Test
    public void write() throws IOException
    {
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        final EmptyUnion emptyUnion = new EmptyUnion();
        emptyUnion.write(writer);
        byte bytes[] = writer.toByteArray();
        assertEquals(0, bytes.length);
        BitStreamReader reader = new ByteArrayBitStreamReader(bytes);
        EmptyUnion readEmptyUnion = new EmptyUnion(reader);
        assertEquals(emptyUnion, readEmptyUnion);
    }
};
